File: lld/week-02-03-patterns/factory/go/factory_test.go
package factory

import "testing"

func TestCreateEmailNotification(t *testing.T) {
	n, err := CreateNotification(Email)
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	got := n.Send("alice@example.com", "hello")
	want := "Email to alice@example.com: hello"
	if got != want {
		t.Fatalf("expected %q, got %q", want, got)
	}
}

func TestCreateSMSNotification(t *testing.T) {
	n, err := CreateNotification(SMS)
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	got := n.Send("+1-555-0100", "hello")
	want := "SMS to +1-555-0100: hello"
	if got != want {
		t.Fatalf("expected %q, got %q", want, got)
	}
}

func TestCreatePushNotification(t *testing.T) {
	n, err := CreateNotification(Push)
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	got := n.Send("device-123", "hello")
	want := "Push to device-123: hello"
	if got != want {
		t.Fatalf("expected %q, got %q", want, got)
	}
}

func TestCreateUnknownNotificationType(t *testing.T) {
	_, err := CreateNotification(NotificationType(99))
	if err != ErrUnknownNotificationType {
		t.Fatalf("expected ErrUnknownNotificationType, got %v", err)
	}
}

func TestDemo(t *testing.T) {
	for _, nt := range []NotificationType{Email, SMS, Push} {
		n, err := CreateNotification(nt)
		if err != nil {
			t.Fatalf("unexpected err for type %v: %v", nt, err)
		}
		if n.Send("user", "test") == "" {
			t.Fatalf("expected non-empty send result")
		}
	}
}

File: lld/week-02-03-patterns/factory/java/src/FactoryTest.java
/**
 * Plain assert-based test runner (no JUnit dependency needed).
 * Run via `java -cp out Main` (invoked from Main) or `java -cp out FactoryTest` directly.
 */
public class FactoryTest {

    public static void main(String[] args) {
        runAll();
    }

    public static void runAll() {
        testCreateEmailNotification();
        testCreateSMSNotification();
        testCreatePushNotification();
        testCreateUnknownNotificationType();
        System.out.println("All FactoryTest cases passed.");
    }

    private static void testCreateEmailNotification() {
        Notification n = NotificationFactory.create(NotificationType.EMAIL);
        assertEquals("Email to alice@example.com: hello", n.send("alice@example.com", "hello"), "email format");
    }

    private static void testCreateSMSNotification() {
        Notification n = NotificationFactory.create(NotificationType.SMS);
        assertEquals("SMS to +1-555-0100: hello", n.send("+1-555-0100", "hello"), "sms format");
    }

    private static void testCreatePushNotification() {
        Notification n = NotificationFactory.create(NotificationType.PUSH);
        assertEquals("Push to device-123: hello", n.send("device-123", "hello"), "push format");
    }

    private static void testCreateUnknownNotificationType() {
        try {
            NotificationFactory.create(null);
            throw new AssertionError("expected UnknownNotificationTypeException");
        } catch (NotificationFactory.UnknownNotificationTypeException e) {
            // expected
        }
    }

    private static void assertEquals(Object expected, Object actual, String label) {
        if (!expected.equals(actual)) {
            throw new AssertionError(label + ": expected " + expected + " but got " + actual);
        }
    }
}
